src/main/java/edu/usf/cutr/gtfs_realtime/bullrunner/GtfsRealtimeProviderImpl.java

```diff
@@ -219,17 +221,31 @@ private void refreshTripVehicle() throws IOException, JSONException {
             }
 
             // Loop through vehicleArray to build vehiclePosition for the given route
+            HashSet addedVehicleIds = new HashSet(); // keep track of vehicles that have been added to the feed
             for (int k = 0; k < vehicleArray.length(); k++) {
                 JSONObject vehicleObj = vehicleArray.getJSONObject(k);
+                // check if vehicle has been added to feed

```

Comment:
Under what conditions would a duplicate vehicle get added to the feed if we don't use the HashSet?

From our previous discussion I thought this was intended to solve the unknown route ID issue.  But, if we're checking for "unknown" route_ids in the below if statement, do we actually need the HashSet to make sure we're not adding duplicate vehicles to the feed?

Let me know if the duplicate vehicle IDs issue is separate from the unknown route issue.

Reply:
Both issues happen at the same time. The program loop through `route_id`s given in GTFS. Syncromatics data has the same `route_id` for route C and MSC Express. When we can't figure out whether a bus is of route C or MSC Express, it is picked up by both routes and hence duplicated in the feed.

Reply:
So do we need to call `downloadVehiclesRouteC()` twice then, once for each of our GTFS `route_ids`?

It seems like we're duplicating the exact same calls to the server and looping through the same responses twice for external_route_id 156. It seems like it would be simpler to just call `downloadVehiclesRouteC()` once and then within that method call the arrivals APIs, and then create a single `JSONArray` with the route_id for the vehicles set to "C", "MSC Express", or "Unknown".  It would half the number of API calls for "C" and "MSC Express", unless I'm missing something.  This would also let you get rid of the HashSet later to remove the duplicates.

What do you think of the above?

Reply:
Yes that works. That's a little more to code but will reduce run time. I will change the methods that way.

Reply:
Ok, great. I suppose we may still need a Hash Set before calling this method just to keep track if we've already called the downloadVehiclesRouteC() method for the given external_route_id when iterating through the routes from routes.txt. This is an extensible way to set it up so that we could use the same design for the Lot 18 Express later if that's part of Route A, and we add another record to routes.txt later for that route.

Reply:
I used `if (route_id.equals("MSC Express")) continue;` for this idea.

Reply:
@minhhpham I guess that works fine too, and is simpler. :+1: